pkg/network/self_sign_configuration.go

```diff
@@ -0,0 +1,130 @@
+)
+
+// validateSelfSignConfiguration validates the following fields
+// - CARotateInterval
+// - CAOverlapInterval
+// - CertRotateInterval
+func validateSelfSignConfiguration(conf *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.SelfSignConfiguration == nil {
+		return []error{}
+	}
+
+	selfSignConfiguration := *conf.SelfSignConfiguration
+
+	errs := []error{}
+
+	caRotateInterval, err := parseCertificateKnob("caRotateInterval", selfSignConfiguration.CARotateInterval)
+	errs = appendOnError(errs, err)
+
+	caOverlapInterval, err := parseCertificateKnob("caOverlapInterval", selfSignConfiguration.CAOverlapInterval)
+	errs = appendOnError(errs, err)
+
+	certRotateInterval, err := parseCertificateKnob("certRotateInterval", selfSignConfiguration.CertRotateInterval)
+	errs = appendOnError(errs, err)
+
+	// If they cannot be parsed don't continue
+	if len(errs) > 0 {
+		return errs
+	}
+
+	err = validateGreaterThanZero("caRotateInterval", caRotateInterval)
+	errs = appendOnError(errs, err)
+
+	err = validateGreaterThanZero("caOverlapInterval", caOverlapInterval)
+	errs = appendOnError(errs, err)
+
+	err = validateGreaterThanZero("certRotateInterval", certRotateInterval)
+	errs = appendOnError(errs, err)
+
+	// If we have a zero value don't continue
+	if len(errs) > 0 {
+		return errs
+	}
+
+	err = validateGreaterThan("caRotateInterval", caRotateInterval, "caOverlapInterval", caOverlapInterval)
+	errs = appendOnError(errs, err)
+
+	err = validateGreaterThan("caRotateInterval", caRotateInterval, "certRotateInterval", certRotateInterval)
+	errs = appendOnError(errs, err)
+
+	return errs
+}
+
+func fillDefaultsSelfSignConfiguration(conf, previous *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.SelfSignConfiguration == nil || conf.SelfSignConfiguration.CARotateInterval == "" || conf.SelfSignConfiguration.CAOverlapInterval == "" || conf.SelfSignConfiguration.CertRotateInterval == "" {
+		if previous != nil && previous.SelfSignConfiguration != nil {
+			conf.SelfSignConfiguration = previous.SelfSignConfiguration
+			return []error{}
+		}
+
+		conf.SelfSignConfiguration = &opv1alpha1.SelfSignConfiguration{
+			CARotateInterval:   caRotateIntervalDefault.String(),
+			CAOverlapInterval:  caOverlapIntervalDefault.String(),
+			CertRotateInterval: certRotateIntervalDefault.String(),
+		}
+
+	}
+	return []error{}
+}
+
+func changeSafeSelfSignConfiguration(prev, next *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if prev.SelfSignConfiguration != nil && next.SelfSignConfiguration != nil && !reflect.DeepEqual(prev.SelfSignConfiguration, next.SelfSignConfiguration) {
+		return []error{errors.Errorf("cannot modify SelfSignConfiguration configuration once it is deployed")}
+	}
+	return []error{}
+}
+
+func parseCertificateKnob(name, value string) (*time.Duration, error) {
+	if value == "" {
+		return nil, nil
+	}
+	d, err := time.ParseDuration(value)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse %s", name)
+	}
+	return &d, nil
```

Comment:
We always have a non-nil duration returned here. Why do you return reference? If we drop the reference, we can get rid of all the nil checks in other functions

Reply:
that's not true

```golang
if value == "" {                                                            
        return nil, nil                                                         
    }    
```

Reply:
And if we change the validation? Empty config (with "" attributes) will not be allowed. It must be `nil` or fully populated. Then we can depend on it being set to a value.